### PR TITLE
[pwrmgr] Prepare d2s countermeasure list and annotations

### PR DESCRIPTION
**Pull request by tjaychen**
Signed-off-by: Timothy Chen <timothytim@google.com>

### CODE REVIEW

File: hw/ip/pwrmgr/doc/_index.md [hunk 1]
```diff
@@ -13,7 +13,8 @@ This document specifies the functionality of the OpenTitan power manager.
 - Software initiated low power entry and hardware requested low power exit.
 - Peripheral reset requests
 - Low power abort and low power fall-through support.
-- ROM integrity check at power-up
+- ROM integrity check at power-up.
+- Local checks for escaltor and power stability.
```
Comment (msfschaffner) [inline hw/ip/pwrmgr/doc/_index.md:17]: nit: escaltor -> escalator?

File: hw/ip/pwrmgr/doc/_index.md [hunk 2]
```diff
@@ -238,26 +245,20 @@ The software is also able to enable recording during `Active` state if it choose
...
+Once flash is powered up and ready, an indication is sent to the [flash_ctrl]({{< relref "hw/top_earlgrey/ip/ast/doc" >}}).
 
-*   Poll flash control register to ensure flash has powered up (this is the same signal mentioned above).
-*   Obtain flash redundancy or configuration information (could be in flash or OTP)
-    *   If in flash, this information is unscrambled
-    *   This item is TBD and must be further discussed.
-*   Program redundancy or configuration information into the flash controller and lock it from further modification (this item is TBD).
-*   Initialize flash scrambling and program any excluded regions based on OTP or other information.
+Once the boot ROM is allowed to execute, it is expected to further initialize the flash prior to using it.
+This involves 4 steps:
```
Comment (msfschaffner) [inline hw/ip/pwrmgr/doc/_index.md:252]: Maybe we should say something like `the following steps`?
It says 4, but there appear to be only 2 bullet points in the list below...

File: hw/ip/pwrmgr/doc/_index.md [hunk 3]
```diff
@@ -303,7 +304,55 @@ There is currently no standardized way to do this, so it is up to the debugging
...
+
+The handling for fall-through and abort are similar to normal sleep exit.
+Since in these scenarios the system was not reset, software continues executing the instruction after the wait-for-interrupt invocation.
+
+1. Check exit condition to determine appropriate steps.
+2. Clear low power hints and configuration in {{< regref "CONTROL" >}}.
+3. Disable wakeup sources and stop recording.
+4. Re-enable interrupts for normal operation.
+
+For an in-depth discussion, please see [power management programmers model](https://docs.google.com/document/d/1w86rmvylJgZVmmQ6Q1YBcCp2VFctkQT3zJ408SJMLPE/edit?usp=sharing) for additional details.
```
Comment (msfschaffner) [inline hw/ip/pwrmgr/doc/_index.md:355]: Is this actually publicly visible? If not, I am not sure whether we should link it publicly...

**Review by msfschaffner — APPROVED** [review]
LGTM, with a few minor questions / nits.

File: hw/ip/pwrmgr/doc/_index.md (hunk 3 above)
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:355]: i think eventually it will be made public.

File: hw/ip/pwrmgr/doc/_index.md (hunk 2 above)
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:252]: ah yes good point i removed a few but didn't update the number.
i'll switch to just steps like you suggest.

**Review by matutem — CHANGES_REQUESTED** [review]

File: hw/ip/pwrmgr/doc/_index.md [hunk 4]
```diff
@@ -238,26 +245,20 @@ The software is also able to enable recording during `Active` state if it choose
...
 ### Power-Down Handling
 
-When the device enters deep sleep, the flash is automatically put into power down mode by the AST.
-When the device exits low power state, it is the responsibility of the boot ROM to poll for flash power-up complete.
-
+Before the device enters low power, the pwrmgr first checks to ensure there are no ongoing transactions to the flash storage.
+When the device enters deep sleep, the flash storage is automatically put into power down mode by the AST.
+When the device exits low power state, it is the responsibility of the boot ROM to poll for flash power-up complete similar to the above section.
 
 ### Flash Brownout Handling
```
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:263]: sounds good. Did you mean to highlight the section below?

File: hw/ip/pwrmgr/doc/_index.md [hunk 5]
```diff
@@ -303,7 +304,55 @@ There is currently no standardized way to do this, so it is up to the debugging
...
+Assume first the system has the power states described [above](#supported-low-power-modes).
+
+## Programmer Sequence for Entering Low Power
+
+1. Disable interrupts
+2. Enable desired wakeup and reset sources in {{< regref "WAKEUP_EN" >}} and {{< regref "RESET_EN" >}}.
+3. Perform any system-specific low power entry steps, e.g.
+   - Interrupt checks (if something became pending prior to disable)
+4. Configure low power mode in {{< regref "CONTROL" >}}.
+5. Set low power hint in {{< regref "LOW_POWER_HINT" >}}.
```
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:319]: o yes good point.

File: hw/ip/pwrmgr/doc/_index.md [hunk 6]
```diff
@@ -303,7 +304,55 @@ There is currently no standardized way to do this, so it is up to the debugging
...
+6. Execute wait-for-interrupt instruction on the processing host.
+
+### Possible Exits
+
+Once low power is entered, the system may exit due to several reasons.
+1. Graceful low power exit - This exit occurs when some source in the system gracefully wakes up the power manager.
+2. [Fall through exit](#fall-through-handling) - This exit occurs when an interrupt manages to break the wait-for-interrupt loop.
+3. [Aborted entry](#abort-handling) - This exit occurs when low power entry is attempted with an ongoing non-volatile transaction.
```
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:327]: wait sorry, what do you mean by some reset?

[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:327]: oooo wait i see what you mean.

File: hw/ip/pwrmgr/doc/_index.md [hunk 7]
```diff
@@ -303,7 +304,55 @@ There is currently no standardized way to do this, so it is up to the debugging
...
+### Exiting from Deep Sleep
+
+When exiting from deep sleep, the system begins execution in ROM.
+
+1. Complete normal preparation steps.
+2. Check reset cause in [rstmgr]({{< relref "hw/ip/rstmgr/doc" >}})
+3. Re-enable modules that have powered down.
+4. Disable wakeup recording through {{< regref "WAKE_INFO_CAPTURE_DIS" >}}.
+5. Check which source woke up the system through {{< regref "WAKE_INFO" >}}.
+6. Take appropriate steps to handle the wake and resume normal operation.
```
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:350]: added

File: hw/ip/pwrmgr/doc/_index.md [hunk 8]
```diff
@@ -303,7 +304,55 @@ There is currently no standardized way to do this, so it is up to the debugging
...
+
+### Exiting from Normal Sleep
+
+The handling for fall-through and abort are similar to normal sleep exit.
+Since in these scenarios the system was not reset, software continues executing the instruction after the wait-for-interrupt invocation.
+
+1. Check exit condition to determine appropriate steps.
+2. Clear low power hints and configuration in {{< regref "CONTROL" >}}.
+3. Disable wakeup sources and stop recording.
```
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md]: added

File: hw/ip/pwrmgr/doc/_index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:267]: let me know if the added language and clarifications here regarding `flash storage` and `flash_ctrl` makes sense to you.

[the remark this answers is not in the document]
Comment (matutem) [reply hw/ip/pwrmgr/doc/_index.md:267]: What I wanted to know is whether the AST will lower por_n_i. You say the response from AST is indistinguishable from POR, so I assume the answer is yes.

**Review by matutem — APPROVED** [review]
LGTM

File: hw/ip/pwrmgr/doc/_index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/ip/pwrmgr/doc/_index.md:267]: let me try to make this a bit clearer. `por_n_i` is actually an input into ast and not an output.

**Comment by tjaychen** [comment]
let me merge this first and i will add some more text to explain that other part.

**Pull request merged**